### PR TITLE
Better thread safety for JsonClaimSet Claims and JsonWebToken Audiences

### LINKED ISSUE #2186: [Bug] Better thread safety for JsonClaimSet Claims and JsonWebToken Audiences

### PR DESCRIPTION
**Pull request by keegan-caruso**
#2186

### CODE REVIEW

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs [hunk 1]
```diff
@@ -650,28 +651,35 @@ public IEnumerable<string> Audiences
...
-                    }
-#else
-                    if (Payload.TryGetValue(JwtRegisteredClaimNames.Aud, out JsonElement audiences))
+                    lock (_audiencesLock)
                     {
-                        if (audiences.ValueKind == JsonValueKind.String)
-                            _audiences = new List<string> { audiences.GetString() };
-
-                        if (audiences.ValueKind == JsonValueKind.Array)
+                        if (_audiences == null)
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs:656]: do you need to check again for null? see line 652

File: src/Microsoft.IdentityModel.JsonWebTokens/Json/JsonClaimSet45.cs [hunk 2]
```diff
@@ -122,35 +123,46 @@ internal IList<Claim> CreateClaims(string issuer)
...
-                    _claims.Add(new Claim(entry.Key, claimValue, ClaimValueTypes.String, issuer, issuer));
-                    continue;
-                }
-
-                var jtoken = entry.Value;
-                if (jtoken != null)
-                {
-                    AddClaimsFromJToken(_claims, entry.Key, jtoken, issuer);
-                    continue;
+                    if (_claims == null)
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.JsonWebTokens/Json/JsonClaimSet45.cs:130]: two null checks? Are they needed?

**Review by jennyf19 — APPROVED** [review]

File: src/Microsoft.IdentityModel.JsonWebTokens/Json/JsonClaimSet45.cs (hunk 2 above)
Comment (keegan-caruso) [reply src/Microsoft.IdentityModel.JsonWebTokens/Json/JsonClaimSet45.cs:130]: yes, this is a standard double checked lock

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs (hunk 1 above)
Comment (keegan-caruso) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs:656]: see other comment, this is a standard double checked lock pattern

Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs:656]: Yes it is necessary, only new up the list if its not null.

**Review by brentschmaltz — APPROVED** [review]

**Pull request merged**